### ISSUE TITLE
[no-missing-imports]: add `tryExtensions` option

### ISSUE THREAD

**Issue by piranna** [eslint-community/eslint-plugin-n#33]
Similar to https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-missing-import.md, add support for the `tryExtensions` option.

**Comment by aladdin-add** [comment]
if using esm, technically it's a bug to use "tryExtensions" - the ext is always required, just like the web.

to support the nodejs esm, we have removed the option in d24be36e7a2ee405b268cd057f92b31558525233.

**Comment by piranna** [comment]
ESM requires to have extensions always, and I agree on that, but Typescript on the other hand forbids them, so without extensions, files are not found...

**Comment by aladdin-add** [comment]
well, we could reintroduce the option to support something like ts.

**Comment by piranna** [comment]
that would be really nice :-) Maybe disabled by default, enforcing extensions to be added.

**Comment by jordanbtucker** [comment]
> ESM requires to have extensions always, and I agree on that, but Typescript on the other hand forbids them, so without extensions, files are not found...

@piranna Can you clarify what you mean? TypeScript allows you to have extensions in your module specifiers because it just copies them verbatim.

```ts
import util from './util.js'
```

transpiles to

```js
const util = require('./util.js')
```

So you shouldn't use `.ts` file extensions in module specifiers like this.

```ts
import util from './util.ts'
```

because it just transpiles to

```js
const util = require('./util.ts')
```

which breaks because Node.js and browsers can't parse TypeScript.

Note that you can still have a `util.ts` file, you just need to import it as if it is a `.js` file that exists relative to the transpiled `.js` file that will be importing it.

**Comment by piranna** [comment]
> @piranna Can you clarify what you mean? TypeScript allows you to have extensions in your module specifiers because it just copies them verbatim.

Yes, but that feels like a nasty hack, because in source code there's no `.js` files, only `.ts` ones, being the `.js` just only on the transpiled code, so this plugin will also fails because there's no `.js` files at all on source files. Correct solution should be Typescript following ESM rules and allow to have `.ts` extensions on imports, but that's not the case. So, the alternative is to have back the `tryExtensions` option, and set it to `.ts` files.

**Comment by jordanbtucker** [comment]
Thanks for clarifying. I agree that adding the `tryExtensions` option for imports will fix the vast majority of use cases for TypeScript users.

**Comment by Lioness100** [comment]
Any progress on this? I'm using `--experimental-specifier-resolution=node` flag, which allows for extensions to be omitted in esm, which is breaking this rule.

**Comment by ashercoren** [comment]
@aladdin-add  Any progress on this?

**Comment by uiolee** [comment]
I'm experiencing the same issue. It is preventing me from migrating from eslint-plugin-node to eslint-plugin-n.
